src/github.com/getlantern/balancer/balancer.go

```diff
@@ -95,20 +75,19 @@ func (b *Balancer) dialerAndConn(network, addr string, targetQOS int) (*Dialer,
 	// To prevent dialing infinitely
 	attempts := 3
 	for i := 0; i < attempts; i++ {
-		if len(dialers) == 0 {
+		if dialers.Len() == 0 {
 			return nil, nil, fmt.Errorf("No dialers left to try on pass %v", i)
 		}
 		var d *dialer
-		d, dialers = randomDialer(dialers, targetQOS)
+		d = heap.Pop(&dialers).(*dialer)
```

Comment:
Specifically, my concern here is that once we find a "best" proxy, as long as that doesn't fail, we'll never try any of the other proxies, so we don't have an opportunity to find out if any of them has gotten better than the one that we thought was "best".